client/report.go

```diff
@@ -0,0 +1,95 @@
+package client
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+)
+
+const reportEndpoint = "/transactions.xml"
+
+//ReportAppID - Report for the Application Id authentication pattern with serviceToken
+func (client *ThreeScaleClient) ReportAppID(appId string, token string, tokenType string, serviceId string, transactions ReportTransactions) (ApiResponse, error) {
+	values := parseQueries(transactions, url.Values{}, transactions.Usage, transactions.Log)
+
+	switch tokenType {
+	case "service_token":
+		values.Add("service_token", token)
+	case "provider_key":
+		values.Add("provider_key", token)
+	default:
+		return ApiResponse{}, errors.New("invalid token type value")
+	}
+
+	values.Add("app_id", appId)
+	values.Add("service_id", serviceId)
+
+	return client.report(values)
+}
+
+//ReportUserKey - Report for the API Key authentication pattern with service token
+func (client *ThreeScaleClient) ReportUserKey(userKey string, token string, tokenType string, serviceId string, transactions ReportTransactions) (ApiResponse, error) {
+	values := parseQueries(transactions, url.Values{}, transactions.Usage, transactions.Log)
+
+	switch tokenType {
+	case "service_token":
+		values.Add("service_token", token)
+	case "provider_key":
+		values.Add("provider_key", token)
+	default:
+		return ApiResponse{}, errors.New("invalid token type value")
+	}
+
+	values.Add("user_key", userKey)
+	values.Add("service_id", serviceId)
+
+	return client.report(values)
+}
+
+func (client *ThreeScaleClient) report(values url.Values) (ApiResponse, error) {
+	var resp ApiResponse
+
+	req, err := client.buildGetReq(reportEndpoint)
+	if err != nil {
+		return resp, errors.New(httpReqError.Error() + " for AuthRep")
+	}
+
+	req.URL.RawQuery = values.Encode()
+	resp, err = client.doHttpReq(req)
+	if err != nil {
+		return resp, fmt.Errorf("error calling 3Scale API - %s", err.Error())
+	}
+	return resp, nil
+}
+
+type ReportTransactions struct {
+	AppID     string  `query:"app_id"`
+	UserKey   string  `query:"user_key"`
+	UserId    string  `query:"user_id"`
+	Timestamp string  `query:"timestamp"`
+	Usage     Metrics `query:"usage"`
```

Comment:
This is not going to work right now due the seralization the client is doing.. I will fix it, this is just to discuss if this way of doing things is better or not.